File: connected/fcnetwork.py
#! fcnetwork.py
# A fully connected feed-forward neural network from scratch 
# (allowing for numpy & random).  Gradient descent for quadratic
# cost function minimization, the option of ReLu or sigmoid 
# activation functions, trained on individual training examples

### TODO: network only learns with sigmoid activation function: fix
### relu design to test this function as well.

import random
import numpy as np 

# Nielsen's mnist dataset loader
import mnist_loader

class FullNetwork:

	def __init__(self, architecture, activation_function):

		# initialize biases
		self.biases = []
		for layer in architecture[1:]:
			self.biases.append(np.random.randn(layer, 1))
		self.arc = architecture

		self.activation = activation_function

		# initialize weights
		self.weights = []
		for i in range(1, len(architecture)):
			j = architecture[i]
			k = architecture[i-1]
			self.weights.append(np.random.randn(k, j))

		self.item_length = len(architecture)


	def activation_function(self, z):
		# Activation function, rectified linear unit
		z = np.array(z)
		zero_array = np.zeros(z.shape)

		if self.activation == 'relu':
			return np.maximum(zero_array, z)

		if self.activation == 'sigmoid':
			return 1 / (1 + np.exp(-z))


	def activation_prime(self, z):
		# derivative of ReLu or sigmoid expression

		if self.activation == 'relu':
			return np.where(z > 0, 0, 1)

		if self.activation == 'sigmoid':
			return self.activation_function(z)*(1 - self.activation_function(z))


	def network_output(self, output):
		# Feed-forward network output
		for index in range(self.item_length-1):
			weight, bias = self.weights[index], self.biases[index]
			output = self.activation_function(np.dot(weight.T, output) + bias)

		return output


	def evaluate(self, test_data):
		test_results = [(np.argmax(self.network_output(x)), np.argmax(y)) for (x, y) in test_data]
		return sum(int(x == y) for (x, y) in test_results)


	def cost_function_derivative(self, output_activations, y):
		# quadratic cost function derivative
		return output_activations - y


	def gradient_descent(self, data, epochs, learning_rate, bootstrap=False):
		# Trains neural network using stochastic gradient descent 
		# with optional bootstrapping.
		data = list(data)
		for epoch in range(epochs):
			for entry in data[:40000]:
				self.update_network(entry, learning_rate)
			number_correct = self.evaluate(data[40000:50000])
			number_total = len(data[40000:50000])
			print ('Epoch {} complete: {} / {}'.format(epoch, number_correct, number_total))


	def update_network(self, entry, learning_rate):
		_input, classification = entry

		# set input activation
		activation = _input
		activations = [activation]
		z_vectors = []

		# compute subsequent layer activations and z vectors
		for i in range(1, self.item_length):
			z_vector = np.dot(self.weights[i-1].T, activation) + self.biases[i-1]
			activation = self.activation_function(z_vector)
			activations.append(activation)
			z_vectors.append(z_vector)

		# compute output error
		error = self.cost_function_derivative(activations[-1], classification) * self.activation_prime(z_vectors[-1])

		# Find partial derivatives of the last layer wrt error
		dc_db = []
		dc_dw = []
		dc_db.append(error)
		dc_dw.append(np.dot(error, activations[-2].T))

		# backpropegate to previous layers
		for i in range(2, self.item_length):
			error = np.dot(self.weights[-i+1], error) * self.activation_prime(z_vectors[-i])

			# update partial derivatives with new error
			dc_db.append(error)
			dc_dw.append(np.dot(error, activations[-i-1].T))

		# update weights and biases
		dc_db.reverse()
		dc_dw.reverse()

		dc_dw = np.array(dc_dw)

		partial_db = [dnb for dnb in dc_db]
		partial_dw = [dnw.T for dnw in dc_dw]

		# gradient descent (move the opposite direction of the gradient)
		lr = learning_rate
		self.weights = [w - lr*dw for w, dw in zip(self.weights, partial_dw)]
		self.biases = [b - lr*db for b, db in zip(self.biases, partial_db)]



training_data, validation_data, test_data = mnist_loader.load_data_wrapper()

net = FullNetwork([784, 20, 10], activation_function='sigmoid')

net.gradient_descent(training_data, 20, 0.1, bootstrap=False) # data, epochs, learning_rate, bootstrap

File: connected/fc_network.cpp

// fc_network.cpp
/* 
 * A neural network for MNIST digit classification. Sigmoid activation
 * function, stochastic gradient descent
 * 
 * */

#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <random>

using namespace std;

float random_sample(float mean, float stdev){
	std::random_device rd; 
    std::mt19937 gen(rd()); 
    
    float sample;
	// instance of class std::normal_distribution with specific mean and stddev
	normal_distribution<float> distro(mean, stdev); 

	// get random number with normal distribution using gen as random source
	sample = distro(gen); 
	
	return sample;
	}

class Network{
	vector<int> architecture = {784, 20, 10};
	int network_length = architecture.size();
	vector<vector<float>> biases;
	vector<vector<vector<float>>> weights;

	 // initialize biases
	public: 
		void biases_init(){
			for (int i=0; i < network_length-1; i++){
				vector<float> temp;
				for (int j=0; j < architecture[i]; j++){
					 //obtain number from normal distribution
					float val = random_sample(0, 1); 
					temp.push_back(val);
				}
				biases.push_back(temp);
			}
			return;
		}
	
		//void print_biases(){
			//cout << biases.size();
			//for (auto u: biases){
				//for (auto v: u){
					//cout << v;
				//}
			//}
		//}
	
	 //initialize weights
		void weights_init(){
			for (int i=0; i < network_length-1; i++){
				vector<vector<float>> layer;
				for (int j=0; j < architecture[i]; j++){
					vector<float> temp;
					for (int k=0; k < architecture[i]; k++){
						float val = random_sample(0, 1);
						temp.push_back(val);
					}
					layer.push_back(temp);
				}
				weights.push_back(layer);
			}
			return;
		}
	
		//void print_weights(){
			//for (auto u: weights){
				//for (auto v: u){
					//for (auto q:v){
						//cout << q;
					//}
				//}
			//}
		//}
	
	float activation_function(float z){
		float sigmoid_z = 1 / (1+pow(2.7828, -z));
		return sigmoid_z;
	};
	
	float activation_prime(float z){
		float sigmoid_prime_z = activation_function(z) * (1 - activation_function(z));
		return sigmoid_prime_z;
	};
	
	float cost_function_derivative(float output_activations, float y){
		return output_activations - y;
	};
	
	
	vector<float> network_output(vector<float> input_arr){
		//feed forward output from neural network from reformatted input
		vector<float> z_matrix;
		
		for (int i=0; i < network_length-1; i++){
			float weight = weights[i];
			float bias = biases[i];
			vector<float> total_ls = [0]*bias.size();
			for (int j=0; j < weight.size(); j++){
				float w_vec = weight[j];
				for (int k=0; k < w_vec.size(); k++){
					total_ls += w_vec[k] * input_arr[j];
				}
			}
			
		}
		
		return z_matrix;
	};
	

};



//train the network 
int main() {
	Network connected_net;
	connected_net.biases_init();
	connected_net.weights_init();
	//connected_net.print_weights();
	//connected_net.gradient_descent(10, 50, 0.1);
	return 0;
	}















